### PR TITLE
[Z-Way] Binding patch to prevent slowing down over time; Issue #3106

### PR DESCRIPTION
**Pull request by Wire82**
[Z-Way] Binding patch to prevent slowing down over time

The patch prevents that the binding slows down over time. 
With the current state of the binding the binding start to slow down and command handling from OpenHab to zWay takes more and more time.
To prevent that this happens the use "ZWay-library-for-Java" was improved and updated. This includes the way the library handles the http client generation.
Also some part of the binding are change to improved the latency of the binding.
The normal handling of the binding doesn't change by the patch
The issue is discussed within community thread:
https://community.openhab.org/t/new-z-way-binding/14096/270
This patch solves issue #3106 

An actual .jar file for testing  is available here:
https://github.com/Wire82/openhab2-addons/releases

The change of the binding is tested by the community here:
https://community.openhab.org/t/new-z-way-binding/14096/269?u=wire82

### CODE REVIEW

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java [hunk 1]
```diff
@@ -634,6 +636,7 @@ public void handleCommand(ChannelUID channelUID, final Command command) {
                 refreshChannel(channel);
             }
         }
+        logger.debug("Handeling of command took {} milliseconds", System.currentTimeMillis() - startTime);
```
Comment (renestraub) [inline addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java]: Potential typo: should be Handling

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java [hunk 2]
```diff
@@ -466,14 +467,15 @@ private void zwayUnsubscribeOpenHabItem(Item openHABItem) {
 
     @Override
     public void handleUpdate(ChannelUID channelUID, State newState) {
+        // Deprecated, therefore deactivated
         // Refresh update time
-        logger.debug("Handle update for channel: {} with new state: {}", channelUID.getId(), newState.toString());
-
-        refreshLastUpdate();
+        // logger.debug("Handle update for channel: {} with new state: {}", channelUID.getId(), newState.toString());
+        // refreshLastUpdate();
```
Comment (renestraub) [inline addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java]: Do we really want to have commented out Code in here?

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java (hunk 1 above)
Comment (Wire82) [reply addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java]: Typo corrected

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java (hunk 2 above)
Comment (Wire82) [reply addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java]: Agree. Code removed

**Comment by Wire82** [comment]
@renestraub Thank you for the review! I changed your remarks accordingly.

**Comment by Wire82** [comment]
@kaikreuzer 
When do you think you can review and hopflully approve the PR? As @renestraub already mentioned within the community the change makes a huge different for many users between a usless installation and a well working system. 
With the current unchanged binding the Z-Way binding can’t be used. 
Thanks Wire82

File: addons/binding/org.openhab.binding.zway/META-INF/MANIFEST.MF [hunk 3]
```diff
@@ -17,7 +17,7 @@ Import-Package:
  com.google.common.collect,
  com.google.gson,
  com.google.gson.annotations,
- com.google.gson.internal,
+ com.google.gson.internal;resolution:=optional,
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.zway/META-INF/MANIFEST.MF]: Why this change?

Comment (Wire82) [reply addons/binding/org.openhab.binding.zway/META-INF/MANIFEST.MF]: If you want to install the binding via the addon folder you get an unsolved dependency and you have to change it to get the binding running. I also noticed that at some other bindings. It can be that it’s not a problem if it is integrated into openHAB. But by using the .jar file this change is needed.

Comment (martinvw) [reply addons/binding/org.openhab.binding.zway/META-INF/MANIFEST.MF]: But then you can install it have a class not found is soon as it loading any class with an import  to `com.google.gson.internal`

**Comment by htreu** [comment]
> this binding becomes dependant on the discovery to get the device list I think that is the wrong direction

@martinvw & @Wire82, I agree it looks strange to push the device list from the discovery service to the bridge handler. My naive assumption is that the bridge handler, or even better the device handler itself caches an instance of `Device` it will handle. This way only the first access will be using the `zwayBridgeHandler.getZWayApi().getDevices()` call. Later calls should check if the cached `Device` still matches the `deviceId`.
wdyt?

File: addons/binding/org.openhab.binding.zway/META-INF/MANIFEST.MF (hunk 3 above)
Comment (Wire82) [reply addons/binding/org.openhab.binding.zway/META-INF/MANIFEST.MF]: Hmm I haven’t seen any issue with an class trying to import com.google.gson.internal. 
Can you give me another hint why this can be a problem.

**Comment by Wire82** [comment]
@htreu & @martinvw 
I think I understand your concerns. At the moment the deviceList is stored within the bridge handler and only updated by the discovery. Is this relay an issue?
I wanted keep the change of the binding as small as possible.

**Comment by htreu** [comment]
The thing is, you will not be able to control your devices in case discovery is not running (for whatever reason). Its best to keep discovery and thing handling separate.
Don't worry about the size of your change set. The idea to cache the communication instance is a good one and you can make any change which turns this into the better.

**Comment by Wire82** [comment]
@htreu @martinvw 
I agree on most points. But I think that controlling the device is also possible if device discovery is not working because the `bridgeHandler` is holding the latest `deviceList`. 
An idea is to get the `deviceList` by the `bridgeHandler` during initialization and only get an update by the discovery. This would ensure that during startup the `bridgeHandler` gets the latest `deviceList` independent from discovery. 
If you would add a new device you need to have a functional device discovery and it should be a reliable way to update the `deviceList` within the `bridgeHandler`.
But storing the `Device` within the `deviceHandler` is also a way. I think that the `Device` should not change over time. The `deviceID` isn't changing because it is stored within the properties. And the id should return the same device every time.
tdyt?

File: addons/binding/org.openhab.binding.zway/META-INF/MANIFEST.MF (hunk 3 above)
Comment (Wire82) [reply addons/binding/org.openhab.binding.zway/META-INF/MANIFEST.MF]: So I checked again, and starting from eclipse is working fine without optional. But for the single .jar file it is required or the binding wouldn't start.
So I think it could be removed for the PR.
tdyt?

**Comment by htreu** [comment]
> I think that controlling the device is also possible if device discovery is not working

It's not about the possibility of control loss, but the dependency between discovery and the ThingHandler. Thing discovery and thing handling must not be dependent. Thus a ThingHandler is supposed to hold a communication object to independently communicate with the device (in this case the deviceList in the bridge).

Why is the call to `zwayBridgeHandler.getZWayApi().getDevices()` not made in the `initialise` method and again when a new thing handler is created?

**Comment by Wire82** [comment]
> Why is the call to zwayBridgeHandler.getZWayApi().getDevices() not made in the initialise method and again when a new thing handler is created?

Good idea. I changed the binding accordingly. The `bridgeHandler` now updates the `deviceList` during initialization and again if an `childHandler` is initialized or disposed

File: addons/binding/org.openhab.binding.zway/META-INF/MANIFEST.MF (hunk 3 above)
Comment (Wire82) [reply addons/binding/org.openhab.binding.zway/META-INF/MANIFEST.MF]: I removed the optional for the PR. From the Eclipse it works fine without it

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java [hunk 4]
```diff
@@ -466,14 +467,12 @@ private void zwayUnsubscribeOpenHabItem(Item openHABItem) {
...
-        // Refresh update time
-        logger.debug("Handle update for channel: {} with new state: {}", channelUID.getId(), newState.toString());
-
-        refreshLastUpdate();
+        // Deprecated, therefore deactivated        
     }
 
     @Override
     public void handleCommand(ChannelUID channelUID, final Command command) {
+        long startTime = System.currentTimeMillis();
```
Comment (htreu) [inline addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java]: I guess you used the time taken here and debug log to show that your change really improved the speed. Is it still needed now? Otherwise I suggest to remove it again.

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java [hunk 5]
```diff
@@ -634,6 +633,7 @@ public void handleCommand(ChannelUID channelUID, final Command command) {
                 refreshChannel(channel);
             }
         }
+        logger.debug("Handling of command took {} milliseconds", System.currentTimeMillis() - startTime);
```
Comment (htreu) [inline addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java]: see comment above.

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayBridgeHandler.java [hunk 6]
```diff
@@ -107,6 +110,8 @@ public void run() {
                 if (mZWayApi.getLogin() != null) {
                     // Thing status set to online in login callback
                     logger.info("Z-Way bridge successfully authenticated");
+                    // Gets the latest deviceList from zWay during bridge initialization
+                    deviceList = mZWayApi.getDevices();
```
Comment (htreu) [inline addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayBridgeHandler.java]: Yes, thats the right way to do it! Nice.

Comment (Wire82) [reply addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayBridgeHandler.java]: Perfect. Thanks. :)

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java (hunk 4 above)
Comment (Wire82) [reply addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java]: Yes this was the purpose.
From my side it is okay to remove it but maybe it can also be helpful to identify the reason for a possible latency during command handling.
Or do you think it is not precise enough for that purpose?

Comment (htreu) [reply addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java]: It might be precise enough but imho it should not be in "production" code. I leave it up to you but my suggestion is to remove it.

Comment (Wire82) [reply addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java]: Yes you are right. I removed it.

**Comment by Wire82** [comment]
@martinvw and @htreu
After all points are incorporated or changed, I hope that the binding now is ready to be approved. 
wdyt?

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayBridgeHandler.java [hunk 7]
```diff
@@ -490,6 +495,32 @@ protected ZWayBridgeConfiguration getZWayBridgeConfiguration() {
         return mConfig;
     }
 
+    /*******************************
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayBridgeHandler.java]: I personally dislike banners :-)

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayBridgeHandler.java [hunk 8]
```diff
@@ -570,6 +601,14 @@ public IZWayApi getZWayApi() {
         return mZWayApi;
     }
 
+    public DeviceList getDeviceList() {
+        return deviceList;
+    }
+
+    public void setDeviceList(DeviceList deviceList) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayBridgeHandler.java]: This does not seem be used, I would prefer not to have it

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayBridgeHandler.java (hunk 7 above)
Comment (martinvw) [reply addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayBridgeHandler.java]: But I see that we used it before

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java [hunk 9]
```diff
@@ -466,10 +467,7 @@ private void zwayUnsubscribeOpenHabItem(Item openHABItem) {
 
     @Override
     public void handleUpdate(ChannelUID channelUID, State newState) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java]: Should we maybe just remove the whole method then from the class?

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java [hunk 10]
```diff
@@ -68,6 +68,7 @@
  * sent to one of the channels.
  *
  * @author Patrick Hecker - Initial contribution
+ * @author Johannes Einig - handleUpdate deactivated (deprecated); handleCommand debug info added
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayDeviceHandler.java]: I would not say that the following description explains the core of your changes:

> handleUpdate deactivated (deprecated); handleCommand debug info added

Maybe mention the caching

File: addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayBridgeHandler.java (hunk 7 above)
Comment (Wire82) [reply addons/binding/org.openhab.binding.zway/src/main/java/org/openhab/binding/zway/handler/ZWayBridgeHandler.java]: Okay. Left it. All other points have been incorporated

**Comment by martinvw** [comment]
Thank you very much for the changes and your patience with us :-)

**Comment by Wire82** [comment]
@martinvw 
Thank you also for your comment. :-)

**Comment by Wire82** [comment]
@martinvw 
Sorry one last questions and I don't know if this is the right place, but I recognize an error with the Discovery Service on the Z-Way binding.
Is there a known issue with that? I haven't changed anything by this PR

**Comment by martinvw** [comment]
What is:

> but I recognize an error with the Discovery Service on the Z-Way binding.

An error log, a stacktrace, a message in the paperUI?

**Comment by Wire82** [comment]
Oh sorry it's a stacktrace message. I send you the details via Gitter
